### DISCUSSION TITLE
Zarr / OME-Zarr support

### DISCUSSION THREAD

**Discussion by jstriebel** [libvips/libvips#3516] [Ideas]
It would be great if libvips could support [Zarr](https://zarr.dev), and possibly also [OME-Zarr](https://github.com/ome/ngff) (also called OME-NGFF) for metadata. Zarr is a chunked storage format and API specification for n-dimensional arrays, and a common format in the bio-imaging and geo-spatial domains, especially for large datasets. Metadata is written in json files, and chunks are stored as separate files by default, optionally compressed by different algorithms, such as gzip or blosc. It is often used in cloud contexts, but also with local files. For libvips, support of local files would be awesome, I'm not sure if and how support of remote storage such as GCP or AWS would fit in.

Zarr just approved the [v3 specification](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html), which probably should be considered for implementation instead of v2.

I'm aware of the following C / C++ libraries for Zarr:
* https://github.com/constantinpape/z5
  (A pure C API is discussed in https://github.com/constantinpape/z5/issues/68)
* https://github.com/google/tensorstore
* https://github.com/xtensor-stack/xtensor-zarr
* https://github.com/Unidata/netcdf-c

OME-Zarr (or OME-NGFF) is the successor to OME-Tiff, which in either case adds a metadata specification and layout structure for microscopy images. Supporting this is not super important IMO, but might be useful for storing relevant metadata, such as resolutions.

It would be great and interesting to hear some thoughts of the maintainers about adding Zarr support. About me: I'm a co-author of the Zarr v3 spec, and happy user of libvips.

**Comment by jcupitt** [comment]
Hi @jstriebel,

Sure, this sounds interesting. libvips allows C++ APIs in dependencies, so I'd think z5 would work.

Let's make an enhancement issue for this. 

A tiny proof of concept PR so we can experiment with it would probably be the next step.

**Comment by jcupitt** [comment]
Further discussion should probably be in the enhancement issue: https://github.com/libvips/libvips/issues/3517
